File: chapter_4/code_4.6.py
# From Bayesian Models for Astrophysical Data 
# by Hilbe, de Souza & Ishida, 2017, Cambridge Univ. Press

# Code 4.7 - Multivariate normal linear model in Python using STAN
# 1 response (y) and 2 explanatory variables (x1, x2)

import numpy as np
import statsmodels.api as sm
import pystan

from scipy.stats import uniform, norm

# Data
np.random.seed(1056)                 # set seed to replicate example
nobs= 5000                           # number of obs in model 
x1 = uniform.rvs(size=nobs)          # random uniform variable
x2 = uniform.rvs(size=nobs)          # second explanatory

X = np.column_stack((x1,x2))        # create response matrix
X = sm.add_constant(X)              # add intercept
beta = [2.0, 3.0, -2.5]             # create vector of parameters

xb = np.dot(X, beta)                                  # linear predictor, xb
y = np.random.normal(loc=xb, scale=1.0, size=nobs)    # create y as adjusted
                                                      # random normal variate 
# Fit
toy_data = {}                                # build data dictionary
toy_data['nobs'] = nobs                      # sample size
toy_data['x'] = X                            # explanatory variable         
toy_data['y'] = y                            # response variable
toy_data['k'] = toy_data['x'].shape[1]       # number of explanatory variables

# STAN code
stan_code = """
data {
    int<lower=1> k;  
    int<lower=0> nobs;                                 
    matrix[nobs, k] x;                     
    vector[nobs] y;                     
}
parameters {
    matrix[k,1] beta;                                             
    real<lower=0> sigma;               
}
transformed parameters{
    matrix[nobs,1] mu;
    vector[nobs] mu2;

    mu = x * beta;
    mu2 = to_vector(mu);                 # normal distribution 
                                          # does not take matrices as input
}
model {
    for (i in 1:k){                       # Diffuse normal priors for predictors
        beta[i] ~ normal(0.0, 100);
    }
    sigma ~ uniform(0, 100);            # Uniform prior for standard deviation

    y ~ normal(mu2, sigma);               # Likelihood function
}
"""

# Run mcmc
fit = pystan.stan(model_code=stan_code, data=toy_data, iter=5000, chains=3,
                  n_jobs=3, verbose=False)

# Output
nlines = 9                                   # number of lines in screen output

output = str(fit).split('\n')
for item in output[:nlines]:
    print(item)   

File: chapter_4/code_4.6.R
# From: Bayesian Models for Astrophysical Data, Cambridge Univ. Press
# (c) 2017,  Joseph M. Hilbe, Rafael S. de Souza and Emille E. O. Ishida 
# 
# you are kindly asked to include the complete citation if you used this 
# material in a publication

# Code 4.6 -cMultivariate normal linear model in R using JAGS

require(R2jags)

set.seed(1056)                              # set seed to replicate example
nobs = 5000                                 # number of obs in model

x1 <- runif(nobs)                           # random uniform variable
x2 <- runif(nobs)                           # random uniform variable
beta1 = 2.0                                 # intercept
beta2 = 3.0                                 # 1st coefficient
beta3 = -2.5                                # 2nd coefficient

xb <- beta1 + beta2*x1 + beta3*x2           # linear predictor
y <- rnorm(nobs, xb, sd=1)                  # create y as adjusted random normal variate

# Model setup
X <- model.matrix(~ 1 + x1+x2)
K <- ncol(X)

model.data <- list(Y = y,                  # response variable
                   X = X,                  # predictors
                   K = K,                  # number of predictors including the intercept
                   N = nobs                # sample size
                   )

NORM <- "
model{
    # Diffuse normal priors for predictors
    for (i in 1:K) { beta[i] ~ dnorm(0, 0.0001) }

    # Uniform prior for standard deviation
    tau <- pow(sigma, -2) # precision
    sigma ~ dunif(0, 100) # standard deviation

    # Likelihood function
    for (i in 1:N){
        Y[i] ~ dnorm(mu[i],tau)
        mu[i] <- eta[i]
        eta[i] <- inprod(beta[], X[i,])
    }
}"

inits <- function () {
  list (
    beta = rnorm(K, 0, 0.01))
}

params <- c ("beta", "sigma")

normOfit <- jags(data = model.data,
                  inits = inits,
                  parameters = params,
                  model = textConnection(NORM),
                  n.chains = 3,
                  n.iter = 15000,
                  n.thin = 1,
                  n.burnin = 10000)

print (normOfit, intervals=c(0.025, 0.975), digits=2)